scripts/install-cni.sh

```diff
@@ -131,19 +131,119 @@ else
 fi
 
 # Fill CNI spec template.
-ipv4_subnet=$(jq '.spec.podCIDR' <<<"$response")
+#######################################
+# Checks if given subnet is valid IPv4 range.
+# Arguments:
+#   Subnet
+# Returns:
+#   0 if valid, 1 on invalid.
+#######################################
+function is_ipv4_range {
+  local IPV4_VALIDATION_REGEXP='^([0-9]{1,3}\.){3}[0-9]{1,3}\/[0-9]{1,2}$'
+  local ip=$1
+
+  if [[ "${ip:-}" =~ ${IPV4_VALIDATION_REGEXP} ]] ; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+#######################################
+# Checks if given subnet is valid IPv6 range.
+# Arguments:
+#   Subnet
+# Returns:
+#   0 if valid, 1 on invalid.
+#######################################
+function is_ipv6_range {
+  local IPV6_VALIDATION_REGEXP='^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))\/[0-9]{1,3}$'
+  local ip=$1
+
+  if [[ "${ip:-}" =~ ${IPV6_VALIDATION_REGEXP} ]] ; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+#######################################
+# Replaces `@subnets` and `@routes` placeholders using `.spec.podCIDRs` from node json.
+# Arguments:
+#   cni spec template
+#   node json from kube-apiserver
+# Outputs:
+#   template with substituted placeholders
+#######################################
+function fillSubnetsInCniSpec {
+  local cni_spec=$1
+  local node=$2
+  local node_ipv6_addr=$3
+
+  local SUBNETS_REPLACEMENT=""
+  local ROUTES_REPLACEMENT=""
+
+  for subnet in $(jq -r '.spec.podCIDRs[]' <<<"$node") ; do
+    if is_ipv4_range "$subnet" ; then
+      if [ "" != "${SUBNETS_REPLACEMENT}" ] ; then
+        SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, "
+        ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, "
+      fi
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}[{\"subnet\": \"${subnet}\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}{\"dst\": \"0.0.0.0/0\"}"
+    elif is_ipv6_range "$subnet" ; then
+      if [ "" != "${SUBNETS_REPLACEMENT}" ] ; then
+        SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, "
+        ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, "
+      fi
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}[{\"subnet\": \"${subnet}\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}{\"dst\": \"::/0\"}"
+    else
+      echo "[ERROR] Subnet $subnet is not a valid IP range"
+      exit 1
+    fi
+  done
+
+  # Directpath usecase
```

Comment:
Are you sure this is only for directpath case today?

Reply:
I'm not 100% sure, but the condition looks like this is used when either `ENABLE_IPV6` or `DUAL_STACK` is in place. `DUAL_STACK` has IPv6 subnet defined in k8s node object in `.spec.podCIDRs` so dual stack doesn't need this code path. So I'm assuming it's only needed when `ENABLE_IPV6` is in place and I'm reading the `ENABLE_IPV6` flag as "directpath".

Actually I found that ENABLE_IPV6 flag is set when either directpath or dualstack are in place:
```
"EnableIPv6":                directpathEnabled || network.IsDualStackCluster(o.Cluster),
```

So for dual-stack this code path should be specifically omited to prevent duplicate subnet entries. I'll adjust testcases to cover that.